### DISCUSSION TITLE
PowerBuilder 2021 fork of the PFC?

### DISCUSSION THREAD

**Discussion by doedewaldt** [OpenSourcePFCLibraries/2019#33] [General]
Bruce are you planning on creating a fork for PowerBuilder 2021 of the PFCLibraries?  I am converting my applications from Powerbuilder 11.5.1 to PowerBuilder 2021 and one of them I wrote ages ago was PFC Based.  I have some of my own additions that I would like to see if they would add value to others that I added to the PFC ages ago.

EDIT:  I may not add anything - I have to see if what I have added is not already handled some other way as there is a log of new functionality compared to my PFC that is version 5.0....

An introduction - I have been programming in PowerBuilder since sometime towards the end of 1992 or beginning of 1993 and have and have a couple of applications that have been in production since 1994.  I have just been adding to and managing them since then.

**Comment by bruce-armstrong** [comment]
At some point.  However, the portion of code that was version specific was refactored to determine the runtime version dynamically.  So unless we make modifications that reference controls or functions in later release that aren't available in an earlier release the code should work on any 2019 or later version of PowerBuilder.  I've actually been using PowerBuilder 2021 to make most of my recent changes.

See https://github.com/OpenSourcePFCLibraries/2019/commit/a4c0cf31e9305b6e4650ec4fe53f5143ff73b19d

The bigger problem comes when Appeon adds something in the middle of a version.  For example, the RibbonBar control was added in 2019 R2.  When we add wrappers for new controls like that, we'll break compatibility with earlier releases of the same major version.

**Reply by bruce-armstrong** [reply]
Actually, I take that back.  The code is dynamic, but the CHOOSE CASE is still hard coded for certain values.  We'll need to add values there for 2021.
